### Import modules

In this question we only need numpy module:

In [36]:
import numpy as np

### Part 1: Generating 80 random numbers in range of [-5000, 5000)

we can use `np.random.uniform()` function to generate random float numbers with uniform distribution:

In [37]:
rand_numbers = np.random.uniform(low = -5000, high = 5000, size = 80)
rand_numbers

array([ 4261.05024587, -2446.49278108, -2133.9655781 , -4691.64419539,
       -1379.18327056,  2482.49894745, -4053.61182446,  4157.52876885,
       -4003.81313232, -3201.32736184,  1267.05777083,  1576.04340277,
        3395.63040937, -4540.59683396, -2825.48434921,   207.99811333,
       -3667.06769056,  1731.65650768,  4672.09180032,  1945.02042289,
        2327.06291665,  3215.66826749,  4890.13776707,  4789.43705719,
        2236.92326055, -1900.45238551, -3955.27352627,  1875.38798945,
         704.2795152 , -1674.21515658,  3038.56982393,   872.39094132,
        4167.37152821, -1246.9168465 ,  1661.16820743,  1426.5652133 ,
       -4216.33383218,   168.85924108, -2579.55239676,  2523.83114466,
         146.69659942, -4620.20017469, -4748.8700833 , -4590.55611995,
        3917.91523556, -1845.19926171, -4419.92114578, -4442.18187432,
        3395.85842859,  1045.51925139, -3789.34977915, -1183.63848113,
        -800.86234121, -2341.59322387,   172.75764474, -2974.03056446,
      

### Part 2: Checking output type and element types

We can use python built-in `type()` function to check output type:

In [38]:
print(f'Type of the output: {type(rand_numbers)}')

Type of the output: <class 'numpy.ndarray'>


Since out output is a `numpy.ndarray` we can check its element types using `numpy.ndarray.dtype` attribute:

In [39]:
print(f'Type of elements: {rand_numbers.dtype}')

Type of elements: float64


we can see that the output type is `float64`.

### Part 3: Rounding the numbers to the nearest integer number

In order to round the numbers, we can use `numpy.round()` function:

In [40]:
rounded_numbers = np.round(rand_numbers)
rounded_numbers

array([ 4261., -2446., -2134., -4692., -1379.,  2482., -4054.,  4158.,
       -4004., -3201.,  1267.,  1576.,  3396., -4541., -2825.,   208.,
       -3667.,  1732.,  4672.,  1945.,  2327.,  3216.,  4890.,  4789.,
        2237., -1900., -3955.,  1875.,   704., -1674.,  3039.,   872.,
        4167., -1247.,  1661.,  1427., -4216.,   169., -2580.,  2524.,
         147., -4620., -4749., -4591.,  3918., -1845., -4420., -4442.,
        3396.,  1046., -3789., -1184.,  -801., -2342.,   173., -2974.,
       -4866.,   374.,   266.,  -602.,   761.,  1172., -1824.,  2855.,
       -2379.,  -924., -1368.,  3406.,   373.,  4679., -3381.,  1249.,
        1243.,   433.,   682., -4006.,  -410.,  3189.,  3791.,  -143.])

### Part 4: Choosing a suitable data type

we should choose from a vast amount of available data types. Each data type supports a range of numbers considering the type of the number which can be an integer or a float number:

- `int`: As same as `int64`.
- `int8`: Integer number in range of $[-2^7, 2^7 - 1]$
- `uint8`: Integer number in range of $[0, 2^8 - 1]$
- `int16`: Integer number in range of $[-2^{15}, 2^{15} - 1]$
- `uint16`: Integer number in range of $[0, 2^{15} - 1]$
- `int32`: Integer number in range of $[-2^{31}, 2^{31} - 1]$
- `int64`: Integer number in range of $[-2^{63}, 2^{63} - 1]$
- `float`: As same as `float64`.
- `float32`: Floating point number in range of $[-2^{31}, 2^{31} - 1]$
- `float64`: Floating point number in range of $[-2^{63}, 2^{63} - 1]$

In our case, because we've rounded our numbers, we need to support integer numbers in range of [-5000, 5000). So, the best choice is `int16` because it is the smallest data type which can support this range of numbers.

We see that after conversion all of the numbers remain the same as before:

In [41]:
converted_numbers = rounded_numbers.astype(np.int16)
np.all(converted_numbers == rounded_numbers)

True

To prove the correctness of our conversion we can check the data type again:

In [42]:
converted_numbers.dtype

dtype('int16')

### Part 5: Normalizing the numbers

We can use this formula to normalize our numbers and scale them into the range of [0, 1]:  
$x = \frac{x - x_{max}}{x_{max} - x_{min}}$

so we can use this conversion followed by a multiplication to 255 in order to scale our numbers into the range of [0, 255].

In [43]:
ranged_numbers = (converted_numbers - (-5000)) / (5000 - (-5000)) * 255
ranged_numbers

array([236.1555,  65.127 ,  73.083 ,   7.854 ,  92.3355, 190.791 ,
        24.123 , 233.529 ,  25.398 ,  45.8745, 159.8085, 167.688 ,
       214.098 ,  11.7045,  55.4625, 132.804 ,  33.9915, 171.666 ,
       246.636 , 177.0975, 186.8385, 209.508 , 252.195 , 249.6195,
       184.5435,  79.05  ,  26.6475, 175.3125, 145.452 ,  84.813 ,
       204.9945, 149.736 , 233.7585,  95.7015, 169.8555, 163.8885,
        19.992 , 131.8095,  61.71  , 191.862 , 131.2485,   9.69  ,
         6.4005,  10.4295, 227.409 ,  80.4525,  14.79  ,  14.229 ,
       214.098 , 154.173 ,  30.8805,  97.308 , 107.0745,  67.779 ,
       131.9115,  51.663 ,   3.417 , 137.037 , 134.283 , 112.149 ,
       146.9055, 157.386 ,  80.988 , 200.3025,  66.8355, 103.938 ,
        92.616 , 214.353 , 137.0115, 246.8145,  41.2845, 159.3495,
       159.1965, 138.5415, 144.891 ,  25.347 , 117.045 , 208.8195,
       224.1705, 123.8535])

because we have floating point numbers, we can use `numpy.round()` function to round these numbers. after that we can use `numpy.ndarray.astype()` function to convert our numbers to `np.int16`:

In [44]:
ranged_numbers = np.round(ranged_numbers).astype(np.int16)
ranged_numbers

array([236,  65,  73,   8,  92, 191,  24, 234,  25,  46, 160, 168, 214,
        12,  55, 133,  34, 172, 247, 177, 187, 210, 252, 250, 185,  79,
        27, 175, 145,  85, 205, 150, 234,  96, 170, 164,  20, 132,  62,
       192, 131,  10,   6,  10, 227,  80,  15,  14, 214, 154,  31,  97,
       107,  68, 132,  52,   3, 137, 134, 112, 147, 157,  81, 200,  67,
       104,  93, 214, 137, 247,  41, 159, 159, 139, 145,  25, 117, 209,
       224, 124], dtype=int16)

### Part 6: Reshaping the array to (8, 10)

We can use `nummpy.ndarray.reshape()` function to reshape our array:

In [45]:
reshaped_numbers = ranged_numbers.reshape((8, 10))
reshaped_numbers

array([[236,  65,  73,   8,  92, 191,  24, 234,  25,  46],
       [160, 168, 214,  12,  55, 133,  34, 172, 247, 177],
       [187, 210, 252, 250, 185,  79,  27, 175, 145,  85],
       [205, 150, 234,  96, 170, 164,  20, 132,  62, 192],
       [131,  10,   6,  10, 227,  80,  15,  14, 214, 154],
       [ 31,  97, 107,  68, 132,  52,   3, 137, 134, 112],
       [147, 157,  81, 200,  67, 104,  93, 214, 137, 247],
       [ 41, 159, 159, 139, 145,  25, 117, 209, 224, 124]], dtype=int16)

### Part 7: Converting data type to int8

Using `numpy.ndarray.astype()` function we get the array below:

In [47]:
int8_numbers = reshaped_numbers.astype(np.int8)
int8_numbers

array([[ -20,   65,   73,    8,   92,  -65,   24,  -22,   25,   46],
       [ -96,  -88,  -42,   12,   55, -123,   34,  -84,   -9,  -79],
       [ -69,  -46,   -4,   -6,  -71,   79,   27,  -81, -111,   85],
       [ -51, -106,  -22,   96,  -86,  -92,   20, -124,   62,  -64],
       [-125,   10,    6,   10,  -29,   80,   15,   14,  -42, -102],
       [  31,   97,  107,   68, -124,   52,    3, -119, -122,  112],
       [-109,  -99,   81,  -56,   67,  104,   93,  -42, -119,   -9],
       [  41,  -97,  -97, -117, -111,   25,  117,  -47,  -32,  124]],
      dtype=int8)

To check the equality of the new array and previous array:

In [49]:
np.all(reshaped_numbers == int8_numbers)

False

So, it seems that the new array is not equal to the previous one. Lets check in which elements we have these changes:

In [50]:
reshaped_numbers[np.where(reshaped_numbers != int8_numbers)]

array([236, 191, 234, 160, 168, 214, 133, 172, 247, 177, 187, 210, 252,
       250, 185, 175, 145, 205, 150, 234, 170, 164, 132, 192, 131, 227,
       214, 154, 132, 137, 134, 147, 157, 200, 214, 137, 247, 159, 159,
       139, 145, 209, 224], dtype=int16)

unchanged numbers:

In [51]:
reshaped_numbers[np.where(reshaped_numbers == int8_numbers)]

array([ 65,  73,   8,  92,  24,  25,  46,  12,  55,  34,  79,  27,  85,
        96,  20,  62,  10,   6,  10,  80,  15,  14,  31,  97, 107,  68,
        52,   3, 112,  81,  67, 104,  93,  41,  25, 117, 124], dtype=int16)

As I mentioned above, the supported range of `int8` data type is $[-2^7, 2^7 - 1]$. 

In our previous matrix we had numbers in range of $[0, 2^8 - 1]$. So, the new data type does not support the numbers in range of $[2^7, 2^8 - 1]$ and overflow will happen for the numbers in this range.

Thats why we see that all of the changed numbers are more than 127 and all of the unchanged numbers are less than 128.

For the numbers bigger than 127, the new value can be calculated as below:

$x = x - 256$

for example if you look at the first element of the array which has the value of 236:

$236 - 256 = -20$

which is the value of the first element in `int8` array.
